### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by mayankchauhan971**
Recreated ERB version from scratch, previous version supported at-most 2 participants.
- Added resume addition.
resumeUploadParticipantActiveStorage

- Separate Interviewer and Interviewee so that only resumes of interviewee are visible.
resumeUpload
resumeUpload

- Added validation before interview creation to avoid clash of interviews for any participant.
validation

- Added email reminder.
843267d4-fd1d-412b-b44e-2b2745a8f2a4

- Added interview updation email. Also added deleteInterviewMailJob.rb job to avoid sending email reminder of interviews which were rescheduled.
54d64483-7243-4681-95c9-5d34f98f3996

### CODE REVIEW

File: app/controllers/interviews_controller.rb [hunk 1]
```diff
@@ -17,46 +17,53 @@ def edit
...
-    # @interview = Interview.create(interview_params)
-    # allEmails.each do |email|
-      # call the mailer
-    # end
+    if (Interview.clash(interview_params, interview_params[:pemail]))
+      respond_to do |format|
+        format.html { redirect_to interviews_path, notice: 'Couldn,t create interview because of clash.' }
+      end
+    else
+      @interview = Interview.new
```
Comment (jaindheerain) [inline app/controllers/interviews_controller.rb:25]: what is the role of this line if you are creating interview in line 30.

File: app/controllers/interviews_controller.rb [hunk 2]
```diff
@@ -17,46 +17,53 @@ def edit
...
+        # link participant and email for every interview
+        curParticipant = Participant.find_or_create_by(email: email)
+        @interviewParticipants = InterviewParticipant.create(:interview=> @interview, :participant => curParticipant, :position => role)
+        # for testing
+        # InterviewMailer.reminder_mail(@interview, email).deliver_now
+        InterviewMailer.reminder_mail(@interview).deliver_later(wait_until: @interview.start - 30.minutes)
+      end
+
+      respond_to do |format|
+        if @interview.save
```
Comment (jaindheerain) [inline app/controllers/interviews_controller.rb:42]: you used @interview.create, so save has no significance here, your interview has already been created, you should check @interview.present?

File: app/controllers/interviews_controller.rb [hunk 3]
```diff
@@ -17,46 +17,53 @@ def edit
...
-        format.html { render :new }
-        format.json { render json: @interview.errors, status: :unprocessable_entity }
+      allEmails.zip(allRoles).each do |email, role|
+        # link participant and email for every interview
+        curParticipant = Participant.find_or_create_by(email: email)
+        @interviewParticipants = InterviewParticipant.create(:interview=> @interview, :participant => curParticipant, :position => role)
+        # for testing
+        # InterviewMailer.reminder_mail(@interview, email).deliver_now
+        InterviewMailer.reminder_mail(@interview).deliver_later(wait_until: @interview.start - 30.minutes)
+      end
```
Comment (jaindheerain) [inline app/controllers/interviews_controller.rb:39]: this whole interview creation logic should be encapsulated in model.

File: app/controllers/interviews_controller.rb [hunk 4]
```diff
@@ -17,46 +17,53 @@ def edit
...
+    @interview.update(:start => interview_params[:start], :end => interview_params[:end], :title => interview_params[:title])
+    emails = interview_params[:pemail]
+    allEmails = emails.split(",")
+    allEmails.each do |email|
+      # delete previous job
+      DeleteInterviewMailJob.perform_now(prevInterview.id)
+      # create new job
+      InterviewMailer.updation_mail(@interview, email).deliver_now
+      InterviewMailer.reminder_mail(@interview).deliver_later(wait_until: @interview.start - 30.minutes)
     end
```
Comment (jaindheerain) [inline app/controllers/interviews_controller.rb:64]: this whole interview update logic with deletion should be encapsulated in the model.

File: app/jobs/delete_interview_mail_job.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+require 'sidekiq/api'
+class DeleteInterviewMailJob < ApplicationJob
+  queue_as :default
+  def perform(interview_id)
```
Comment (jaindheerain) [inline app/jobs/delete_interview_mail_job.rb:4]: this method is not recommended, please feel free to discuss some other approach

File: app/mailers/application_mailer.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: ENV['GMAIL_EMAIL']
```
Comment (jaindheerain) [inline app/mailers/application_mailer.rb:2]: nice

File: app/models/interview.rb [hunk 7]
```diff
@@ -1,4 +1,35 @@
...
+      end
+    end
+    return false
+  end
+
+  def self.clash(interview_params, emails)
+    # check clash for each participant
+    all_emails = emails.split(",")
+    all_emails.each do |email|
+      participant = Participant.where(email: email).first
```
Comment (jaindheerain) [inline app/models/interview.rb:24]: use .find or find_by

File: app/views/interviews/index.html.erb [hunk 8]
```diff
@@ -5,29 +5,33 @@
...
-        <td><%= interview.date %></td>
-        <td><%= interview.start %></td>
-        <td><%= interview.end %></td>
-        <td><%= interview.title %></td>
-        <td><%= link_to 'Show', interview %></td>
-        <td><%= link_to 'Edit', edit_interview_path(interview) %></td>
-        <td><%= link_to 'Destroy', interview, method: :delete, data: { confirm: 'Are you sure?' } %></td>
-      </tr>
+    <tr>
+      <td><%= interview.start.utc %></td>
```
Comment (jaindheerain) [inline app/views/interviews/index.html.erb:19]: you should show time to the user in user time zone,

File: app/views/interviews/show.html.erb [hunk 9]
```diff
@@ -5,23 +5,30 @@
...
 
 <div>
   <strong>Participant's Email</strong>
-  <% @participants.each do |p| %>
-    <%= p.email %>
-    <br>
+  <% @interview.participants.each do |p| %>
+  <p>
+    <%= p.email %> || 
+    <% curPosition = InterviewParticipant.where(interview_id: @interview.id, participant_id: p.id)%>
```
Comment (jaindheerain) [inline app/views/interviews/show.html.erb:22]: query in a loop is bad.

File: app/views/interviews/show.html.erb [hunk 10]
```diff
@@ -5,23 +5,30 @@
...
   <strong>End:</strong>
   <%= @interview.end %>
 </p>
 
 <div>
   <strong>Participant's Email</strong>
-  <% @participants.each do |p| %>
-    <%= p.email %>
-    <br>
+  <% @interview.participants.each do |p| %>
```
Comment (jaindheerain) [inline app/views/interviews/show.html.erb:19]: this line fetches participants, this is direct interaction with DB in view, this is a bad practice. not recommended.

File: app/controllers/interviews_controller.rb (hunk 1 above)
Comment (mayankchauhan971) [reply app/controllers/interviews_controller.rb:25]: Sir, I noted this thing in my last app that I was getting an error on creating an interview without creating an instance of it. Turns out something else was causing it, I have updated it.

File: app/controllers/interviews_controller.rb (hunk 2 above)
Comment (mayankchauhan971) [reply app/controllers/interviews_controller.rb:42]: Okay, this was generated on scaffolding so didn't bother much, I have read about .present and will incorporate it.

File: app/views/interviews/index.html.erb (hunk 8 above)
Comment (mayankchauhan971) [reply app/views/interviews/index.html.erb:19]: Got it.

File: app/models/interview.rb (hunk 7 above)
Comment (mayankchauhan971) [reply app/models/interview.rb:24]: Okay got it sir, will update.

File: app/controllers/interviews_controller.rb (hunk 4 above)
Comment (mayankchauhan971) [reply app/controllers/interviews_controller.rb:64]: Okay got it sir.

**Pull request merged**